### DISCUSSION TITLE
Jedis cluster pipeline sync problem

### DISCUSSION THREAD

**Discussion by rahul-narayana** [xetorthio/jedis#2991] [Q&A] [answered]
I am experimenting with the new cluster pipeline support in jedis. Here is a sample script in scala

```scala
val jedisClientConfig: DefaultJedisClientConfig = DefaultJedisClientConfig.builder().build()
val clusterConnectionProvider = new ClusterConnectionProvider(
      Set(
          new HostAndPort("h1", 1234),
          new HostAndPort("h2", 5678)
      ).asJava, jedisClientConfig)
var pipeline = new ClusterPipeline(clusterConnectionProvider)
```
As I am not using any pool I guess it is taking the default pool of size 8. I am syncing the pipeline after 10 commands. 

```scala
pipeline.sync()
pipeline = new ClusterPipeline(clusterConnectionProvider)
```

After 80 commands, I guess the pool is exhausted and code freezes to move further. What is the best approach to handle such scenarios. Do I need to use `new ClusterPipeline` after every sync instead of using a `ClusterConnectionProvider`.

`new ClusterPipeline` approach makes a new connection for every sync which will slow down overall performance (running cluster slots again and again). 

Please let me know what would be the ideal approach here. Thanks

**Comment by sazzad16** [comment, marked as answer]
`close()` ClusterPipeline after `sync()`. It'll release the resources (connections) it was using.

```
pipeline.sync()
pipeline.close()
pipeline = new ClusterPipeline(clusterConnectionProvider)
```

**Reply by rahul-narayana** [reply]
Thank you, It worked. The format is different between unclustered and clustered version so I missed this step to close the pipeline
